okta/services/idaas/resource_okta_app_user_custom_schema_property.go

```diff
@@ -238,9 +238,9 @@ func updateAppUserSubSchemaProperty(ctx context.Context, d *schema.ResourceData,
 	retypeUserSchemaPropertyEnums(custom)
 	boc := utils.NewExponentialBackOffWithContext(ctx, 10*time.Second)
 	err = backoff.Retry(func() error {
-		_, resp, err := getOktaClientFromMetadata(meta).UserSchema.
-			UpdateApplicationUserProfile(ctx, d.Get("app_id").(string), *custom)
-		if utils.SuppressErrorOn404(resp, err) == nil {
+		_, resp, err := getOktaClientFromMetadata(meta).UserSchema.UpdateApplicationUserProfile(ctx, d.Get("app_id").(string), *custom)
+		// if error is nil, we skip this check entirely since utils.SuppressErrorOn404 will return nil either when err is nil or when resp.StatusCode is 404.
+		if err != nil && utils.SuppressErrorOn404(resp, err) == nil {
```

Comment:
```suggestion
		if err != nil || utils.SuppressErrorOn404(resp, err) == nil {
```
Should this be an `OR` rather than an `AND`

reasoning
```go
if err != nil && utils.SuppressErrorOn404(resp, err) == nil {
  // this code can never be reached
}

if err != nil || utils.SuppressErrorOn404(resp, err) == nil {
  // this code can be reached if either:
  // 1. err != nil
  // 2. utils.SuppressErrorOn404(resp, err) == nil
}
```

Reply:
With the condition that you're proposing

```
if err != nil || utils.SuppressErrorOn404(resp, err) == nil {
		............
}

```

1. Let's say err != nil  evaluates to true
	since in your case it is logical OR, there is no point in proceeding further, we set an empty ID

2. Let's say err != nil evaluates to false (in other words, there is no error, it's a 2xx status code)
 	since in your case it is logical OR, we will try to evaluate the 2nd part of the condition i.e. `utils.SuppressErrorOn404(resp, err)`
	This is how we call the function --> utils.SuppressErrorOn404(resp, err) 
	Here's the rather small function pasted below for your reference. 

```
// Useful shortcut for suppressing errors from Okta's SDK when a resource does not exist. Usually used during deletion
// of nested resources.
func SuppressErrorOn404(resp *sdk.Response, err error) error {
    if resp != nil && resp.StatusCode == http.StatusNotFound {
        return nil
    }
    return ResponseErr(resp, err)
}
```

If you go through this function substituting nil in place of err, you'll see that eventually it evaluates to nil == nil which is true.
	What we're doing is setting an empty ID when err is in fact nil !

Changing it to && avoids this scenario, we set empty ID only in case of 404.
In case the error is genuine, we do have a block of code further down the line that should take care of it for attributes, finally we return err to what it was set.


If you still believe this is wrong, we can certainly discuss this further but the tests confirm this works.